examples/push-file

```diff
@@ -3,11 +3,64 @@
 import argparse
 import json
 import logging
+import kobo.rpmlib
+import os
 
 from chexus import Client, BucketItem, TableItem
 
 LOG = logging.getLogger("push-file")
 
+def get_sigkey(rpmfile):
+    """
+    If the rpm is signed, return the sigkey.
+    rpmfile: path to an rpm
+    """
+    try:
+        rpmhead = kobo.rpmlib.get_rpm_header(rpmfile)
+    except:
+        raise RuntimeError("Cannot read RPM header: %s" % rpmfile)
+    rpmsig = kobo.rpmlib.get_keys_from_header(rpmhead).lower()
+    if rpmsig != "fd431d51":
```

Comment:
> "fd431d51"
Is **every** RPM really signed with this?

Reply:
Nope, older content and beta content uses different keys.  I think a production publishing tool wouldn't have such a check.

Maybe there's another idea here though, like making it less likely to accidentally use this example to publish pre-release content, which does make sense.

Reply:
Is it feasible to check the rpmsig against a set of all possible keys or otherwise make the check more generic?

Reply:
> older content and beta content uses different keys.

+1. Only 2-3 LPs still use the beta and devel keys to sign new rpms. All other products use the "gold key" (fd431d51) to sign new content (including beta content).

> I think a production publishing tool wouldn't have such a check.

Hm, I would expect it to have this check (not this _exact_ check, but a check to verify the sigkey). I was under the impression that pub contains this sort of rpm signature check at the repo-level -- as in, it checks the rpm's sig against the "gpgkey" field in the repo's content set to determine if the signature is valid for that repo. Maybe that's not the case?

> Is it feasible to check the rpmsig against a set of all possible keys or otherwise make the check more generic?

The list of all possible sigkeys used to sign older content appears to be quite long. For now, I've updated this script to use the 4 sigkeys that I'm familiar with. I assume this would likely cover 99+% of the content. 

> or otherwise make the check more generic?

I don't immediately see how we could make this check more generic. We could check if the rpm contains any signature, but then it could be signed by any "untrusted" gpgkey.


> Maybe there's another idea here though, like making it less likely to accidentally use this example to publish pre-release content, which does make sense.

RHEL (including pre-release alpha and beta content) has been signed with the gold (fd431d51) key for quite some time. After [RCMWORK-20762](https://projects.engineering.redhat.com/browse/RCMWORK-20762) is resolved, and after we've backported old content, I would imagine we should probably remove the other keys and only verify against the gold key.

Reply:
> 
> > I think a production publishing tool wouldn't have such a check.
> 
> Hm, I would expect it to have this check (not this _exact_ check, but a check to verify the sigkey). I was under the impression that pub contains this sort of rpm signature check at the repo-level -- as in, it checks the rpm's sig against the "gpgkey" field in the repo's content set to determine if the signature is valid for that repo. Maybe that's not the case?

The publishing toolchain overall needs such a check, but it ought to happen earlier than this.  If content has been put into a Pulp repo, that's already enough for it to become visible on the customer portal even before we start any "publish to CDN" step, so the wrong key needs to be caught earlier.  You also want to catch the wrong key earlier because that reduces the mess which has to be cleaned up (i.e. removing the wrongly signed RPM from the Pulp repo which attempted to publish it).

So, in theory, having a signing key check also at "publish to CDN" time is:

- unnecessary - because it was already checked earlier in the workflow
- insufficient - because if you only caught the problem at publish time, the wrongly-signed content is already customer visible

This doesn't make a sigkey check at publish time wrong, but it does raise the question whether it's worthwhile.

(If you're thinking "hang on, how can the content already be customer visible before we published it to CDN?" - yeah, there is a design bug here, see https://bugzilla.redhat.com/show_bug.cgi?id=1531376.)

Reply:
> The publishing toolchain overall needs such a check, but it ought to happen earlier than this. If content has been put into a Pulp repo, that's already enough for it to become visible on the customer portal even before we start any "publish to CDN" step, so the wrong key needs to be caught earlier. 

Ah, okay. I was too focused on the example scenario of uploading and publishing a _local_ yum repo directly to aws, and overlooked using the pulp repos as the src repo for upload.

In that case, I agree that this check is not useful here. I'll remove this from the example script to avoid future confusion.